### PR TITLE
Attempt to remove the `where` bounds in arithmetic.

### PR DESCRIPTION
**Pull request by kianenigma**
Needed for https://github.com/paritytech/substrate/pull/7929

So far, these `where` clauses have been haunting me everywhere up to the runtime. The reason for their existence is the use of `<P: PerThing>`, which means that we don't initially know what is the type `P::Inner` and if it meets all the requirements of other functions (like `from_rational_approximation`). 

So far, I have let this be, we need generic `P` in some places. 

During https://github.com/paritytech/substrate/pull/7929, I was quite annoyed by this as the where clause literally propagates to almost every single file and function of staking, if you are to make `OffchainAccuracy` generic. 

Note that we already tolerate and accept this in #7909 as well, but there it is cleaner because everything is wrapper in `impl Pallet`.

So I tried to just remove this where clause from the source, which is this PR. In essence, the only main assumption added is that `type Inner` is always `Into<u128>`. This can easily hold as long as we keep our per-things within the standard primitive types (note: unsigned types are already not allowed, so you have `u8 -> u128` left). But this would disallow using per-things with higher accuracy levels such as U256 etc. 

I had to shuffle a bit of `From<U> for T` into `Into<T> for U`, which to this moment still confuses me, but seemingly it was necessary since rustc does not assume it automatically. At first, I really expected just adding `type Inner: Into<u128>` but it is not.

### CODE REVIEW

File: primitives/arithmetic/src/per_things.rs [hunk 1]
```diff
@@ -37,13 +38,17 @@ pub trait PerThing:
 	Sized + Saturating + Copy + Default + Eq + PartialEq + Ord + PartialOrd + Bounded + fmt::Debug
 {
 	/// The data type used to build this per-thingy.
-	type Inner: BaseArithmetic + Unsigned + Copy + fmt::Debug;
+	type Inner: BaseArithmetic + Unsigned + Copy + Into<u128> + fmt::Debug;
```
Comment (kianenigma) [inline primitives/arithmetic/src/per_things.rs:41]: TLDR ^^^^^^^^^^^^^^^^^^^^^^

File: primitives/arithmetic/src/per_things.rs [hunk 2]
```diff
@@ -91,8 +96,10 @@ pub trait PerThing:
 	/// # }
 	/// ```
 	fn mul_floor<N>(self, b: N) -> N
-	where N: Clone + From<Self::Inner> + UniqueSaturatedInto<Self::Inner> + ops::Rem<N, Output=N> +
-		ops::Div<N, Output=N> + ops::Mul<N, Output=N> + ops::Add<N, Output=N> + Unsigned
+	where
```
Comment (kianenigma) [inline primitives/arithmetic/src/per_things.rs:99]: Sorry for changing the formatting, just did it manual bc the previous one was really not suited for multi where. Rest assured: we only remove the `From<Self::Inner>`

**Comment by thiolliere** [comment]
> I had to shuffle a bit of From<U> for T into Into<T> for U, which to this moment still confuses me, but seemingly it was necessary since rustc does not assume it automatically.

I think this is because: although `From<T> for U` implies `Into<U> for T` the opposite is not true. Into doesn't implies From

**Comment by thiolliere** [comment]
TLDR: we could change `from_rational_approximation` to bound `N: TryFrom<Self::Inner>` instead of `N: From<Self::Inner>`, but it would make code even more tricky for probably no benefit. So I'm fine with the change you proposed.


All this is because `from_rational_approximation` requires `N` to be `From<Self::Inner>`
I required it because we usually used the function with `N` being more than Inner arithmetic.
But now that  `PerThing::Inner` is generic in trait `PerThing` we need to add these where_clause everywhere even when `N` is u128.

But actually if `N` arithmetic fit into `Self::Inner` (i.e. `N` is u16 and `Self::Inner` is u32) then implementing `from_rational_approximation` is doable:
```diff
diff --git a/primitives/arithmetic/src/per_things.rs b/primitives/arithmetic/src/per_things.rs
index c6a31a0ff..cc1c59731 100644
--- a/primitives/arithmetic/src/per_things.rs
+++ b/primitives/arithmetic/src/per_things.rs
@@ -362,7 +362,7 @@ macro_rules! implement_per_thing {
                        }
 
                        fn from_rational_approximation<N>(p: N, q: N) -> Self
-                       where N: Clone + Ord + From<Self::Inner> + TryInto<Self::Inner> + TryInto<Self::Upper>
+                       where N: Clone + Ord + TryFrom<Self::Inner> + TryInto<Self::Inner> + TryInto<Self::Upper>
                                + ops::Div<N, Output=N> + ops::Rem<N, Output=N> + ops::Add<N, Output=N> + Unsigned
                        {
                                let div_ceil = |x: N, f: N| -> N {
@@ -379,6 +379,22 @@ macro_rules! implement_per_thing {
                                // p should not be bigger than q.
                                let p: N = p.min(q.clone());
 
+                               let max: N = if let Ok(max) = $max.try_into() {
+                                       max
+                               } else {
+                                       // $max doesn't fit into `N` arithmetic,
+                                       // thus $max is strictly more than `N::max_value()`,
+                                       // thus `N::max_value()` is strictly less than $max,
+                                       // but p and q are less than `N::max_value()`,
+                                       // thus p and q are less than $max,
+                                       // thus p * $max fits into `Upper` (Upper must be able to compute `$max^2`)
+                                       // and q first into `Inner`
+                                       let part =
+                                               Self::Inner::try_from(p).expect("..") as $upper_type
+                                               * $max as $upper_type
+                                               / q as $upper_type;
+                                       return $name(part as $inner_type)
+                               };
                                let factor: N = div_ceil(q.clone(), $max.into()).max((1 as Self::Inner).into());
 
                                // q cannot overflow: (q / (q/$max)) < $max. p < q hence p also cannot overflow.
```

**Review by thiolliere — APPROVED** [review]

File: primitives/npos-elections/src/lib.rs [hunk 3]
```diff
@@ -332,17 +331,18 @@ impl<AccountId: IdentifierT> Voter<AccountId> {
...
+		let distribution = self
+			.edges
+			.into_iter()
+			.filter_map(|e| {
+				let per_thing = P::from_rational_approximation(e.weight, budget);
 			// trim zero edges.
 			if per_thing.is_zero() { None } else { Some((e.who, per_thing)) }
-		}).collect::<Vec<_>>();
+		})
+			.collect::<Vec<_>>();
```
Comment (shawntabrizi) [inline primitives/npos-elections/src/lib.rs:345]: this formatting cant be right

**Review by shawntabrizi — APPROVED** [review]

File: primitives/npos-elections/src/lib.rs (hunk 3 above)
Comment (kianenigma) [inline primitives/npos-elections/src/lib.rs:345]:
```suggestion
		}).collect::<Vec<_>>();
```

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge
